src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,867 @@
+  toℕ ((suc2* x) + (2suc y))      ≡⟨⟩
+  toℕ (suc (2suc (x + y)))        ≡⟨ toℕ∘suc (2suc (x + y)) ⟩
+  1+ (toℕ (2suc (x + y)))         ≡⟨⟩
+  1+ (2 *ₙ (1+ (toℕ (x + y))))    ≡⟨ cong (1+_ ∘ (2 *ₙ_) ∘ 1+_)
+                                                         (toℕ-homo-+ x y) ⟩
+  1+ (2 *ₙ (1+ (m +ₙ n)))         ≡⟨ cong (1+_ ∘ (2 *ₙ_))
+                                          (Of+ℕ-semigroup.x∙yz≈y∙xz 1 m n)
+                                   ⟩
+  1+ (2 *ₙ (m +ₙ 1+n))            ≡⟨ cong 1+_ (ℕp.*-distribˡ-+ 2 m 1+n) ⟩
+  1+ (2 *ₙ m +ₙ 2 *ₙ 1+n)         ≡⟨ sym (ℕp.+-assoc 1 (2 *ₙ m) (2 *ₙ 1+n)) ⟩
+  (1+ (2 *ₙ m)) +ₙ (2 *ₙ (1+ n))  ≡⟨⟩
+  toℕ (suc2* x) +ₙ toℕ (2suc y)   ∎
+  where
+  m = toℕ x;  n = toℕ y;  1+n = 1+ n
+
+toℕ-homo-+ (suc2* x) (suc2* y) = begin
+  toℕ ((suc2* x) + (suc2* y))       ≡⟨⟩
+  toℕ (suc (suc2* (x + y)))         ≡⟨ toℕ∘suc (suc2* (x + y)) ⟩
+  1+ (toℕ (suc2* (x + y)))          ≡⟨⟩
+  1+ (1+ (2 *ₙ (toℕ (x + y))))      ≡⟨ cong (1+_ ∘ 1+_ ∘ (2 *ₙ_))
+                                                       (toℕ-homo-+ x y) ⟩
+  1+ (1+ (2 *ₙ (m +ₙ n)))           ≡⟨ cong (1+_ ∘ 1+_) (ℕp.*-distribˡ-+ 2 m n)
+                                     ⟩
+  1+ (1+ (2 *ₙ m +ₙ 2 *ₙ n))        ≡⟨ cong 1+_
+                                            (sym (ℕp.+-assoc 1 (2 *ₙ m) _)) ⟩
+  1+ ((1+ (2 *ₙ m) +ₙ (2 *ₙ n)))    ≡⟨ (Of+ℕ-semigroup.x∙yz≈y∙xz
+                                                         1 (1+ (2 *ₙ m)) _) ⟩
+  (1+ (2 *ₙ m)) +ₙ (1+ (2 *ₙ n))    ≡⟨⟩
+  toℕ (suc2* x) +ₙ toℕ (suc2* y)    ∎
+  where
+  m = toℕ x;  n = toℕ y
+
+
+fromℕ-homo-+ :  (m n : ℕ) → fromℕ (m +ₙ n) ≡ fromℕ m + fromℕ n
+fromℕ-homo-+ 0      _ = refl
+fromℕ-homo-+ (1+ m) n = begin
+  fromℕ ((1+ m) +ₙ n)         ≡⟨⟩
+  suc (fromℕ (m +ₙ n))        ≡⟨ cong suc (fromℕ-homo-+ m n) ⟩
+  suc (a + b)                 ≡⟨ sym (suc-+ a b) ⟩
+  (suc a) + b                 ≡⟨⟩
+  (fromℕ (1+ m)) + (fromℕ n)  ∎
+  where
+  a = fromℕ m;  b = fromℕ n
+
+toℕ-injective :  ∀ {x y} → toℕ x ≡ toℕ y → x ≡ y
+toℕ-injective {0#}     {0#}     _               =  refl
+toℕ-injective {2suc x} {2suc y} 2[1+xN]≡2[1+yN] =  cong 2suc x≡y
+  where
+  xN        = toℕ x
+  yN        = toℕ y
+  1+xN≡1+yN = ℕp.*-cancelˡ-≡ {1+ xN} {1+ yN} 1 2[1+xN]≡2[1+yN]
+  xN≡yN     = cong predₙ 1+xN≡1+yN
+  x≡y       = toℕ-injective xN≡yN
+
+toℕ-injective {2suc x} {suc2* y} 2[1+xN]≡1+2yN =
+                               contradiction 2[1+xN]≡1+2yN (even≢odd 1+xN yN)
+  where
+  1+xN = 1+ (toℕ x);  yN = toℕ y
+
+toℕ-injective {suc2* x} {2suc y} 1+2xN≡2[1+yN] =
+                        contradiction (sym 1+2xN≡2[1+yN]) (even≢odd 1+yN xN)
+  where
+  xN = toℕ x;  1+yN = 1+ (toℕ y)
+
+toℕ-injective {suc2* x} {suc2* y} 1+2xN≡1+2yN =  cong suc2* x≡y
+  where
+  xN = toℕ x;  yN = toℕ y;  2xN≡2yN = cong predₙ 1+2xN≡1+2yN
+
+  xN≡yN = ℕp.*-cancelˡ-≡ {xN} {yN} 1 2xN≡2yN;   x≡y = toℕ-injective xN≡yN
+
+
+toℕ-surjective :  (y : ℕ) → ∃ (\x → toℕ x ≡ y)
+toℕ-surjective 0       =  (0# , refl)
```

Comment:
Why not use `fromℕ` here? It would give you a stronger result.

Reply:
Yes, it has changed to
``toℕ-surjective n =  (fromℕ n , toℕ∘fromℕ n)``